<a href="https://colab.research.google.com/github/GiseleDiniz/exemple-colaboratory/blob/main/Desafio_est%C3%A1gio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install requests

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [2]:
import requests
import json
import pandas as pd
from pandas import DataFrame

from requests.auth import HTTPBasicAuth
import numpy as np
import seaborn as sns
import plotly.express as px

import matplotlib.pyplot as plt


In [3]:
url = "https://elasticsearch-saps.saude.gov.br/desc-esus-notifica-estado-*/_search"

In [4]:
payload = json.dumps({
   "size": 1500
})
headers = {
			'Content-Type': 'application/json',
			'Authorization': 'Basic dXNlci1wdWJsaWMtbm90aWZpY2Fjb2VzOlphNHFOWGR5UU5TYTlZYUE='
}

response = requests.request("POST", url, headers=headers, data=payload)
covid = response.json()

In [ ]:
covid['hits']['hits']

In [6]:
df_covid = pd.json_normalize(covid['hits']['hits'])
pd.set_option("display.max_columns", None)
pd.set_option("max_colwidth", None)

In [104]:
df_covid

_index _type         _id  _score _source.codigoRecebeuAntiviral  \
0     desc-esusve_v8  _doc  9tpCbU4BG1     1.0                           None   
1     desc-esusve_v8  _doc  SRPy5KVhNb     1.0                              2   
2     desc-esusve_v8  _doc  mQpxdoptjR     1.0                              3   
3     desc-esusve_v8  _doc  abjJ3WX3kN     1.0                              2   
4     desc-esusve_v8  _doc  yEL4uiHXOu     1.0                           None   
...              ...   ...         ...     ...                            ...   
1495  desc-esusve_v8  _doc  ghvp93BHuW     1.0                              2   
1496  desc-esusve_v8  _doc  hg5JVLVcNF     1.0                              3   
1497  desc-esusve_v8  _doc  zO83TUzJHN     1.0                           None   
1498  desc-esusve_v8  _doc  kYMABwTxzy     1.0                              3   
1499  desc-esusve_v8  _doc  dOCm9no2Tm     1.0                           None   

     _source.outroLocalRealizacaoTestagem _source.municipio  \
0                              DONA LINDU    Rio de Janeiro   
1                                    None      Campo Grande   
2                                    None          Brasília   
3                                    None           Aracaju   
4                                    None           Uberaba   
...                                   ...               ...   
1495                                 None        Santa Rosa   
1496                                 None          Juazeiro   
1497                                 None          Santarém   
1498                                 None         São Paulo   
1499                                 None         São Paulo   

     _source.dataTesteSorologico  _source.registroAtual  _source.idade  \
0                           None                   True             53   
1                           None                   True             67   
2                           None                   True             11   
3                           None                   True             43   
4                           None                   True             37   
...                          ...                    ...            ...   
1495                        None                   True             11   
1496                        None                   True             46   
1497                        None                   True             64   
1498                        None                   True             51   
1499                        None                   True             32   

       _source.dataNotificacao _source.estadoNotificacao  \
0     2023-05-20T03:00:11.826Z                Pernambuco   
1     2023-05-20T04:00:35.854Z        Mato Grosso do Sul   
2     2023-05-20T03:00:45.764Z          Distrito Federal   
3     2023-05-20T03:00:50.225Z                   Sergipe   
4     2022-07-12T03:00:00.000Z              Minas Gerais   
...                        ...                       ...   
1495  2023-05-19T03:00:43.974Z         Rio Grande do Sul   
1496  2022-12-23T03:00:28.783Z                     Bahia   
1497  2023-03-01T03:00:00.000Z                      Pará   
1498  2023-04-19T03:00:15.629Z                 São Paulo   
1499  2022-01-11T03:00:11.559Z                 São Paulo   

     _source.codigoQualAntiviral _source.resultadoTesteSorologicoIgG  \
0                           None                                None   
1                           None                                None   
2                           None                                None   
3                           None                                None   
4                           None                                None   
...                          ...                                 ...   
1495                        None                                None   
1496                        None                                None   
1497                      

In [8]:
covid = response.json()

titles = [item for item in covid['hits']['hits'][2]['_source']]
values = [item for item in covid['hits']['hits'][2]['_source'].values()]


dicionario = {"title": titles, "values": values}

pd.DataFrame(dicionario, columns=['title', 'values'])

title     values
0          codigoRecebeuAntiviral          3
1    outroLocalRealizacaoTestagem       None
2                       municipio   Brasília
3             dataTesteSorologico       None
4                   registroAtual       True
..                            ...        ...
57                           sexo  Masculino
58               recebeuAntiviral   Ignorado
59  laboratorioSegundaReforcoDose       None
60    resultadoTesteSorologicoIgM       None
61         dataSegundaReforcoDose       None

[62 rows x 2 columns]

In [60]:
#pegar a coluna da df
lista_arrumar = list([item['_source']['testes'] for item in covid['hits']['hits']])
lista_geral = list()
for i in lista_arrumar:
  if len(i) > 1:
    lista_resultado = list()
    for j in i:
      teste = pd.DataFrame.from_dict([j])["resultadoTeste"][0]
      lista_resultado.append(teste)
    lista_geral.append(lista_resultado)
  elif len(i) == 1:
    for j in i:
      string_aux = pd.DataFrame.from_dict([j])["resultadoTeste"][0]
      lista_geral.append(string_aux)
  else:
    lista_geral.append("NA")
df_covid["resultadoTeste"] = lista_geral


In [61]:
#excluir a coluna
df_covid.drop(['_source.testes'], axis=1)


_index _type         _id  _score _source.codigoRecebeuAntiviral  \
0     desc-esusve_v8  _doc  9tpCbU4BG1     1.0                           None   
1     desc-esusve_v8  _doc  SRPy5KVhNb     1.0                              2   
2     desc-esusve_v8  _doc  mQpxdoptjR     1.0                              3   
3     desc-esusve_v8  _doc  abjJ3WX3kN     1.0                              2   
4     desc-esusve_v8  _doc  yEL4uiHXOu     1.0                           None   
...              ...   ...         ...     ...                            ...   
1495  desc-esusve_v8  _doc  ghvp93BHuW     1.0                              2   
1496  desc-esusve_v8  _doc  hg5JVLVcNF     1.0                              3   
1497  desc-esusve_v8  _doc  zO83TUzJHN     1.0                           None   
1498  desc-esusve_v8  _doc  kYMABwTxzy     1.0                              3   
1499  desc-esusve_v8  _doc  dOCm9no2Tm     1.0                           None   

     _source.outroLocalRealizacaoTestagem _source.municipio  \
0                              DONA LINDU    Rio de Janeiro   
1                                    None      Campo Grande   
2                                    None          Brasília   
3                                    None           Aracaju   
4                                    None           Uberaba   
...                                   ...               ...   
1495                                 None        Santa Rosa   
1496                                 None          Juazeiro   
1497                                 None          Santarém   
1498                                 None         São Paulo   
1499                                 None         São Paulo   

     _source.dataTesteSorologico  _source.registroAtual  _source.idade  \
0                           None                   True             53   
1                           None                   True             67   
2                           None                   True             11   
3                           None                   True             43   
4                           None                   True             37   
...                          ...                    ...            ...   
1495                        None                   True             11   
1496                        None                   True             46   
1497                        None                   True             64   
1498                        None                   True             51   
1499                        None                   True             32   

       _source.dataNotificacao _source.estadoNotificacao  \
0     2023-05-20T03:00:11.826Z                Pernambuco   
1     2023-05-20T04:00:35.854Z        Mato Grosso do Sul   
2     2023-05-20T03:00:45.764Z          Distrito Federal   
3     2023-05-20T03:00:50.225Z                   Sergipe   
4     2022-07-12T03:00:00.000Z              Minas Gerais   
...                        ...                       ...   
1495  2023-05-19T03:00:43.974Z         Rio Grande do Sul   
1496  2022-12-23T03:00:28.783Z                     Bahia   
1497  2023-03-01T03:00:00.000Z                      Pará   
1498  2023-04-19T03:00:15.629Z                 São Paulo   
1499  2022-01-11T03:00:11.559Z                 São Paulo   

     _source.codigoQualAntiviral _source.resultadoTesteSorologicoIgG  \
0                           None                                None   
1                           None                                None   
2                           None                                None   
3                           None                                None   
4                           None                                None   
...                          ...                                 ...   
1495                        None                                None   
1496                        None                                None   
1497                      

In [62]:
#Qual o estado com a maior proporção de casos confirmados?
df_reagente = df_covid.loc[df_covid['resultadoTeste']== "Reagente"].groupby(['_source.estado'])['resultadoTeste'].count().sort_values().reset_index()
print(df_reagente)


         _source.estado  resultadoTeste
0             Tocantins               1
1               Sergipe               1
2              Rondônia               1
3    Mato Grosso do Sul               1
4               Alagoas               2
5   Rio Grande do Norte               2
6                  Pará               2
7                 Piauí               4
8        Rio de Janeiro               5
9                 Ceará               5
10       Santa Catarina               7
11           Pernambuco               9
12                Goiás              13
13            São Paulo              15
14                 Acre              15
15         Minas Gerais              25
16    Rio Grande do Sul              42


O estado com maior número de casos confirmados é o Rio Grande do Sul, contendo 34 casos confirmados

In [63]:
#Dentre os casos confirmados, qual a proporção de pacientes que receberam ao menos 1 dose de vacina?
def vacina():
  indicador1= df_covid.loc[df_covid['_source.dataPrimeiraDose'] != "None"]
  indicadores2= indicador1.loc[df_covid['resultadoTeste']== "Reagente"]
  valores = indicadores2.groupby(['resultadoTeste'])['_source.dataPrimeiraDose'].count().reset_index()
  
  return valores


In [64]:
vacina()

resultadoTeste  _source.dataPrimeiraDose
0       Reagente                       111

Dentro dos pacientes que testeram postivo para a covid cerca de 111 apenas tomaram a primeira dose de imunizante.

In [100]:
#Como é a distribuição de idade entre pacientes sintomáticos e assintomáticos entre os casos confirmados?
def sintomas():
  indicador1= df_covid.loc[df_covid['_source.sintomas'] != "None"]
  indicadores2= indicador1.loc[df_covid['resultadoTeste']== "Reagente"].groupby(['_source.idade',"_source.sintomas"])['resultadoTeste'].count().reset_index()
  #valores = indicadores2.groupby(['resultadoTeste', '_source.sintomas'])['_source.idade'].count().reset_index()
  return indicadores2

In [101]:
sintomas()

_source.idade                _source.sintomas  resultadoTeste
0                0  Coriza, Tosse, Febre, Dispneia               1
1                4                           Tosse               1
2                6    Coriza, Febre, Dor de Cabeça               1
3                7            Dor de Cabeça, Tosse               1
4               11                   Assintomático               1
..             ...                             ...             ...
140             80                   Assintomático               1
141             83  Coriza, Tosse, Dor de Garganta               1
142             88                 Tosse, Dispneia               1
143             88                   Tosse, Outros               1
144             94     Dor de Cabeça, Tosse, Febre               1

[145 rows x 3 columns]

A distriuição de idade em paciente sintomáticos ou então assintomáticos varia de 0 a 94 anos, ou seja, há a variação de sintomas referente a paciente sintomáticos. 

In [103]:
#Para (pelo menos um) estados a sua escolha, construa uma visualização para acompanhar a evolução dos casos ao longo do período amostrado.
def acompanhamento():
  indicador1= df_covid.loc[df_covid['_source.estado'] == "Pernambuco"]
  indicador2= indicador1.loc[df_covid['resultadoTeste'] == "Reagente"]
  indicador3 = indicador2.loc[df_covid['_source.evolucaoCaso'] != "None"]
  Valormed=indicador3.groupby(['_source.evolucaoCaso', '_source.estado'])['resultadoTeste', '_source.dataNotificacao'].count().reset_index()
  Valormed=Valormed.sort_values(by=["_source.estado","_source.evolucaoCaso"])
  return Valormed
fig = px.bar(acompanhamento(), x="_source.evolucaoCaso", y="resultadoTeste", title='Evolucao dos casos no estado do Rio Grande do Sul')
fig.show()

<ipython-input-103-2d8b535e69b6>:6: FutureWarning:

Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.



In [102]:
indicador2=df_covid.loc[df_covid['resultadoTeste'] == "Reagente"]
Valormed=indicador2.groupby(['_source.evolucaoCaso', '_source.estado'])['resultadoTeste'].count().reset_index()
print(Valormed)

        _source.evolucaoCaso      _source.estado  resultadoTeste
0                  Cancelado               Goiás               6
1                       Cura                Acre              13
2                       Cura             Alagoas               1
3                       Cura               Goiás               5
4                       Cura  Mato Grosso do Sul               1
5                       Cura        Minas Gerais              17
6                       Cura                Pará               2
7                       Cura          Pernambuco               1
8                       Cura               Piauí               1
9                       Cura   Rio Grande do Sul              35
10                      Cura      Rio de Janeiro               4
11                      Cura      Santa Catarina               2
12                      Cura           São Paulo              10
13                      Cura           Tocantins               1
14  Em tratamento domicil

In [22]:
sintomas = [item['_source']['sintomas'] for item in covid['hits']['hits']]

def frequency(sintomas): 
    frequencia = {}
    for item in np.unique(np.array(sintomas)):
         frequencia[item] = np.where(np.array(sintomas)==item)[0].shape[0]

    for key, value in frequencia.items(): 
        print (key, value) 



final_list = frequency(sintomas)
print(final_list)

Assintomático 350
Coriza 17
Coriza, Dispneia 1
Coriza, Dispneia, Outros 1
Coriza, Dispneia, Tosse 2
Coriza, Distúrbios Gustativos, Dispneia 1
Coriza, Distúrbios Gustativos, Distúrbios Olfativos 1
Coriza, Distúrbios Gustativos, Dor de Garganta 1
Coriza, Distúrbios Gustativos, Tosse, Dispneia 2
Coriza, Distúrbios Gustativos, Tosse, Dispneia, Febre 1
Coriza, Distúrbios Gustativos, Tosse, Dor de Cabeça 1
Coriza, Distúrbios Gustativos, Tosse, Dor de Garganta 1
Coriza, Distúrbios Gustativos, Tosse, Febre 1
Coriza, Distúrbios Gustativos, Tosse, Febre, Dor de Garganta, Outros 1
Coriza, Distúrbios Gustativos, Tosse, Febre, Outros, Dor de Garganta 1
Coriza, Distúrbios Olfativos 3
Coriza, Distúrbios Olfativos, Distúrbios Gustativos 1
Coriza, Distúrbios Olfativos, Distúrbios Gustativos, Dor de Cabeça 1
Coriza, Distúrbios Olfativos, Distúrbios Gustativos, Dor de Cabeça, Dor de Garganta, Outros 1
Coriza, Distúrbios Olfativos, Distúrbios Gustativos, Dor de Cabeça, Dor de Garganta, Tosse 1
Coriza, Dis

O sintoma mais presente em pacientes sintomáticos foi a tosse.